### PR TITLE
SpeechLM Data Infra: multimodal IO

### PR DESCRIPTION
**Pull request by jctian98**
@sw005320 

The basic multimodal IO for SpeechLM. 

Prior PRs: https://github.com/espnet/espnet/pull/6257, https://github.com/espnet/espnet/pull/6260

All content is under espent2/speechlm/multimodal_io, which includes:
(1) The abstract class for multimodal IO
(2) The text IO (text.py)
(3) The discrete Audio IO (SSL token, Codec token) and continuous audio IO (pre-trained audio encoder) (audio.py)

This PR mostly works on the initial setup of multimodal IO:
(1) I may need some revision when we connect it to other components later.
(2) It only supports the minimal features used in OpusLM (ESPnet codec, XEUS, HF text tokenizer); but we keep the good interface for future compatibility.
(3) It can support more modalities in the future, like vision.

Follow-up PR in the plan:
(1) SpeechLM preprocessor;
(2) SpeechLM model files
(3) A speechLM-specific trainer
(4) A PR to connect all components
(5) Inference 

Prior PRs: https://github.com/espnet/espnet/pull/6257 https://github.com/espnet/espnet/pull/6260

Request review: @Masao-Someki @wanchichen @siddhu001

### CODE REVIEW

**Comment by sw005320** [comment]
@Masao-Someki, can you review it?

**Comment by sw005320** [comment]
This pull request introduces a multimodal input/output (I/O) abstraction for SpeechLM, implements a HuggingFace-based text I/O class, and adds a utility for padding and batching variable-length sequences. These changes lay the foundation for unified handling of text, audio, and vision modalities, and provide essential utilities for preprocessing and batching data.

**Multimodal I/O Abstraction**
* Added `AbsIO` abstract base class in `espnet2/speechlm/multimodal_io/abs_io.py` to define a unified interface for encoding, decoding, and managing both discrete and continuous representations across modalities (text, audio, vision). This includes methods for batch encoding/decoding, length calculation, feature dimension, stream handling, vocabulary, and stream weighting.

**Text Modality Implementation**
* Implemented `HuggingFaceTextIO` class in `espnet2/speechlm/multimodal_io/text.py`, providing text encoding/decoding using HuggingFace tokenizers. Supports single-stream discrete text, vocabulary retrieval, and stream interval/weight configuration.

**Data Utility**
* Added `pad_list` utility function in `espnet2/speechlm/utils/data.py` for padding and batching variable-length numpy arrays or torch tensors. Handles right-padding, dtype promotion, and returns both padded batch and original lengths.

**Comment by sw005320** [comment]
@wanchichen, please also review this PR

**Comment by jctian98** [comment]
Fix the CI issue and should be ready to review

File: espnet2/speechlm/model/speechlm/multimodal_io/audio.py [hunk 1]
```diff
@@ -0,0 +1,962 @@
+"""Audio I/O implementation for discrete and continuous representations (v2)."""
+
+from pathlib import Path
+from typing import List, Optional, Tuple
+
+import joblib
+import numpy as np
+import torch
+
+from .abs_io import AbsIO
```
Comment (Masao-Someki) [inline espnet2/speechlm/model/speechlm/multimodal_io/audio.py]:
```suggestion
from espnet2.speechlm.model.speechlm.multimodal_io.abs_io import AbsIO
```

File: espnet2/speechlm/model/speechlm/multimodal_io/audio.py [hunk 2]
```diff
@@ -0,0 +1,962 @@
...
+
+
+class KmeansModel(torch.nn.Module):
```
Comment (Masao-Someki) [inline espnet2/speechlm/model/speechlm/multimodal_io/audio.py:18]: I just want to confirm if this class is the same as this class?

https://github.com/espnet/espnet/blob/8c394b2010a48ac36602c9b1ca9b200ec4c20643/egs2/TEMPLATE/asr1/pyscripts/feats/dump_km_label.py#L101C7-L126

File: espnet2/speechlm/multimodal_io/abs_io.py
Comment (Masao-Someki) [inline espnet2/speechlm/multimodal_io/abs_io.py:1]: Is this file the same as `espnet2.speechlm.model.speechlm.multimodal_io.abs_io` ..?
If this is the same file please remove this or `espnet2.speechlm.model.speechlm.multimodal_io`

File: espnet2/speechlm/utils/data.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+"""Data utilities for SpeechLM module."""
+
+from typing import List, Tuple, Union
+
+import numpy as np
+import torch
+
+
+def pad_list(
```
Comment (Masao-Someki) [inline espnet2/speechlm/utils/data.py]: Maybe you can use the existing `pad_list` function and add lengths part? I think it's good for maintenance

https://espnet.github.io/espnet/guide/espnet/nets/pad_list.html

File: espnet2/speechlm/model/speechlm/multimodal_io/audio.py [hunk 4]
```diff
@@ -0,0 +1,962 @@
...
+            )
+            self.ssl_model.eval()
+
+            # Load k-means quantizer with correct device
+            self.km_model = KmeansModel(ssl_kmeans_path, device=self.device)
+
+            # Extract SSL metadata
+            # NOTE: Cannot parse the metadata from hubert_train_args,
+            # using hardcoded values for XEUS
+            self.ssl_n_streams = 1
```
Comment (siddhu001) [inline espnet2/speechlm/model/speechlm/multimodal_io/audio.py]: Again maybe setting these as arguments is better for future development.

File: espnet2/speechlm/model/speechlm/multimodal_io/audio.py (hunk 2 above)
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/multimodal_io/audio.py:18]: it's from there. I made some revision.

File: espnet2/speechlm/multimodal_io/abs_io.py
Comment (jctian98) [reply espnet2/speechlm/multimodal_io/abs_io.py:1]: The old folder `espnet2.speechlm.multimodal_io` will be removed. we only need `espnet2.speechlm.model.speechlm.multimodal_io`

File: espnet2/speechlm/utils/data.py (hunk 3 above)
Comment (jctian98) [reply espnet2/speechlm/utils/data.py]: it's the same function. simply to make `espnet2/speechlm` to be self-consistent

File: espnet2/speechlm/model/speechlm/multimodal_io/audio.py (hunk 1 above)
Comment (jctian98) [reply espnet2/speechlm/model/speechlm/multimodal_io/audio.py]: fixed

File: espnet2/speechlm/model/speechlm/multimodal_io/audio.py (hunk 2 above)
Comment (sw005320) [reply espnet2/speechlm/model/speechlm/multimodal_io/audio.py:18]: Can you just add this as a pointer?
People may confuse the similar functions distributed in various places.
If you add a comment in the header or function explanation with a clear pointer to similar functions, people would not be confused.

Please do so for your entire repo.
I support your self-consistent module design, but we need this care.

File: espnet2/speechlm/utils/data.py (hunk 3 above)
Comment (sw005320) [reply espnet2/speechlm/utils/data.py]: ditto

**Comment by sw005320** [comment]
@jctian98

This PR looks good to me overall.
However, as I mentioned, people may confuse the similar functions distributed in various places.
If you add a comment in the header or function explanation with a clear pointer to similar functions, people would not be confused.

Please do so for your entire module.
I support your self-consistent module design, but we need this care.
